### PR TITLE
ES|QL: add MV_APPEND function

### PR DESCRIPTION
**Pull request by luigidellaquila**
Adding `MV_APPEND(value1, value2)` function, that appends two values creating a single multi-value.
If one or both the inputs are multi-values, the result is the concatenation of all the values, eg.

```
MV_APPEND([a, b], [c, d]) -> [a, b, c, d]
```

~I think for this specific case it makes sense to consider `null` values as empty arrays, so that~
~MV_APPEND(value, null) -> value~
~It is pretty uncommon for ESQL (all the other functions, apart from `COALESCE`, short-circuit to `null` when one of the values is null), so let's discuss this behavior.~

[EDIT] considering the feedback from Andrei, I changed this logic and made it consistent with the other functions: now if one of the parameters is null, the function returns null

### CODE REVIEW

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/meta.csv-spec [hunk 1]
```diff
@@ -227,5 +229,5 @@ countFunctions#[skip:-8.13.99]
 meta functions |  stats  a = count(*), b = count(*), c = count(*) |  mv_expand c;
 
 a:long | b:long | c:long
-99     | 99     | 99
+100    | 100    | 100
```
Comment (luigidellaquila) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/meta.csv-spec:232]: 💯

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Comment by nik9000** [comment]
> I think for this specific case it makes sense to consider `null` values as empty arrays, so that

That's fine - our `null` is basically an empty list. I'd do it.

**Comment by nik9000** [comment]
> It is pretty uncommon for ESQL (all the other functions, apart from `COALESCE`, short-circuit to `null` when one of the values is null), so let's discuss this behavior.

There's a small number that don't - maybe 5 out of the 100.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            "keyword",
+            "long",
+            "text",
+            "version" },
+        description = "Appends values of two multi-value fields."
+    )
+    public MvAppend(
+        Source source,
+        @Param(
+            name = "v1",
```
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:69]: In a discussion in a prior PR(https://github.com/elastic/elasticsearch/pull/106454#pullrequestreview-1945476218), we'd like to give consistent names to function arguments. If an argument takes multiple input data types, for example both string and numeric, we call it field, we can name them field1 and field2 here.

**Review by fang-xing-esql — APPROVED** [review]
LGTM.

**Review by astefan — COMMENTED** [review]
Can you add more csv tests, please? It's just one in there.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+            "double",
+            "geo_point",
+            "geo_shape",
+            "integer",
+            "ip",
+            "keyword",
+            "long",
+            "text",
+            "version" },
+        description = "Appends values of two multi-value fields."
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:65]: Related to the function name, one usually needs to provide the "append-to" target, so even if we were to keep the name, I'd rephrase this to use "Concatenates the..." or "Joins the...".

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppendTests.java [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+
+    @ParametersFactory
+    public static Iterable<Object[]> parameters() {
+        List<TestCaseSupplier> suppliers = new ArrayList<>();
+        booleans(suppliers);
+        ints(suppliers);
+        longs(suppliers);
+        doubles(suppliers);
+        bytesRefs(suppliers);
+        nulls(suppliers);
```
Comment (bpintea) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppendTests.java:46]: Should we cover the rest of the types too? Like, for instance, `MvCount` & co. do?
Slightly off-topic here, it'd be maybe nice to have some tests with other BytesRefs types, like Version and/or IP (in a csv-spec maybe).

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppendTests.java [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        return parameterSuppliersFromTypedData(suppliers);
```
Comment (bpintea) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppendTests.java:47]: ...then we could also chain `parameterSuppliersFromTypedData(errorsForCasesWithoutExamples(anyNullIsNull(...)))`.

**Review by bpintea — COMMENTED** [review]
Left some comments. What jumped first at me was the name, I'd be in favour of finding an alternative.

**Review by astefan — COMMENTED** [review]
`row x = 1 | eval x = mv_append(x)` fails with an unfriendly error message: 

```
                "type": "ql_illegal_argument_exception",
                "reason": "Null children are not allowed",
                "stack_trace": "org.elasticsearch.xpack.ql.QlIllegalArgumentException: Null children are not allowed
```

**Review by astefan — CHANGES_REQUESTED** [review]
From consistency point of view, I would have mv_append return `null` if one of the values is `null`.

Also, `row x = 123, y = null | eval z = mv_append(y, x)` fails with `second argument of [mv_append(y, x)] must be [null], found value [x] type [integer]` while `mv_append(x, y)` is fine with it.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        } else {
+            builder.beginPositionEntry();
+            for (int i = 0; i < count1; i++) {
+                builder.appendInt(field1.getInt(first1 + i));
+            }
+            for (int i = 0; i < count2; i++) {
+                builder.appendInt(field2.getInt(first2 + i));
+            }
+            builder.endPositionEntry();
+        }
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:213]: I'm afraid it doesn't work: if `count1==1` and `count2==1`, `count1 | count2` returns `1`, that means SV, but the final result will be MV (an array of two)

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isType;
+import static org.elasticsearch.xpack.ql.type.DataTypes.NULL;
+
+/**
+ * Appends values to a multi-value
+ */
+public class MvAppend extends ScalarFunction implements OptionalArgument, EvaluatorMapper {
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:46]: The name `mv_append` comes from the [original issue](https://github.com/elastic/elasticsearch/issues/105322) and from the affinity with the same function in other languages.
IMHO "append" makes it clear how the two values will be combined. 
"Join" recalls other concepts; "merge" and "combine" are more generic and ambiguous (eg. what about duplicates?).

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+            "geo_shape",
+            "integer",
+            "ip",
+            "keyword",
+            "long",
+            "text",
+            "version" },
+        description = "Appends values of two multi-value fields."
+    )
+    public MvAppend(
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:67]: This could be a follow-up (it wouldn't break bwc)

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+        this.field1 = field1;
+        this.field2 = field2;
+    }
+
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isType(field1, EsqlDataTypes::isRepresentable, sourceText(), FIRST, "representable");
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:114]: Yeah, I agree... I used "representable" for consistency with the other MV functions, maybe we can review it in general with something more meaningful.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:46]: 👍 fixed

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java (hunk 3 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:65]: 👍 I like "concatenates", much better than "appends"

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+            "date",
+            "double",
+            "geo_point",
+            "geo_shape",
+            "integer",
+            "ip",
+            "keyword",
+            "long",
+            "text",
+            "version" },
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:64]: That would be nice, but I'm afraid the answer is no: we can define a constant `String[]`, but for the annotation also its content has to be constant, and in Java there is no way to guarantee this condition.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppendTests.java (hunk 4 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppendTests.java:46]: `bytesRefs(suppliers)` covers all the types represented as BytesRefs (eg. IP, version, geo), it's just organized differently from MvCount (that has single methods for each type).
+1 to having more CSV tests, I'm adding them.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:213]: Correct, it'd have to be a `switch (count1 + count2) {`

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:46]: > The name mv_append comes from the https://github.com/elastic/elasticsearch/issues/105322 and from the affinity with the same function in other languages.

What I understand with "append" is that the first column is extended with the values of the second one. If I'd see this function in a different language, my first thought would be that the first column is extended and then returned; i.e. after the eval, referencing again the first column (maybe in a subsequent eval) will return an extended list.
But can stay as is too, the docs will cover it.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppendTests.java (hunk 5 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppendTests.java:45]: The validation in this case is a bit peculiar, ie. the supported type for the second parameter (and the error message) depends on the type of the first parameter. I added some custom validation tests.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java [hunk 11]
```diff
@@ -1876,6 +1876,38 @@ public void testInOnText() {
...
+            { "keyword", "keyword" },
+            { "date", "datetime" },
+            { "point", "geo_point" },
+            { "shape", "geo_shape" },
+            { "version", "version" } };
+        int first = randomInt(fields.length - 1);
+        int second = first;
+        while (second == first) {
+            second = randomInt(fields.length - 1);
+        }
```
Comment (bpintea) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:1895]: Optional:
```
Supplier<Int> supplier = () -> randomInt(fields.length - 1);
int first = supplier.get();
int second = randomValueOtherThan(first, supplier);
```

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java [hunk 12]
```diff
@@ -1876,6 +1876,38 @@ public void testInOnText() {
             """, "mapping-multi-field-variation.json", "text");
     }
 
+    public void testMvAppendValidation() {
+        String[][] fields = {
```
Comment (bpintea) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:1880]: +`long`, `ip`? Maybe also `null`?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isType(field1, EsqlDataTypes::isRepresentable, sourceText(), FIRST, "representable");
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        dataType = field1.dataType();
+        return isType(field2, t -> t == dataType, sourceText(), SECOND, dataType.typeName());
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:117]:
```suggestion
        return isType(field2, t -> t == NULL || t == dataType, sourceText(), SECOND, dataType.typeName());
```

Otherwise `MV_APPEND(some_type, null)` will work, but MV_APPEND(null, some_type) won't. Might be worth adding some tests with this too.

**Review by bpintea — APPROVED** [review]
Left some small remarks otherwise LGTM

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java (hunk 13 above)
Comment (bpintea) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:117]: > Checking isType(), it already checks for NULL from what I see: predicate.test(e.dataType()) || e.dataType() == NULL.

In this case the first type, "t" / `dataType`, needs to be checked against `NULL`, rather than `field2`'s type (which is `e` in `isType()`).

>  if any field is null, set dataType to NULL. 

That's correct (and the code does it right also for the `MV_APPEND(null, some_type)`, case). 
But in one case the `null` result has first argument's type, while in the other `null` has the `NULL` type.

**Review by astefan — APPROVED** [review]
LGTM

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Comment by luigidellaquila** [comment]
@elasticmachine update branch
